akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+            tracing: Boolean,
+            role: StreamTls.Role,
+            closing: StreamTls.Closing): Props =
+    Props(new SslTlsCipherActor(settings, sslContext, firstSession, tracing, role, closing))
+
+  final val TransportIn = 0
+  final val TransportOut = 0
+
+  final val UserOut = 1
+  final val UserIn = 1
+}
+
+class SslTlsCipherActor(settings: ActorFlowMaterializerSettings, sslContext: SSLContext,
+                        firstSession: NegotiateNewSession, tracing: Boolean,
+                        role: StreamTls.Role, closing: StreamTls.Closing)
+  extends Actor with ActorLogging with Pump {
+
+  import SslTlsCipherActor._
+
+  protected val outputBunch = new OutputBunch(outputCount = 2, self, this)
+  outputBunch.markAllOutputs()
+
+  protected val inputBunch = new InputBunch(inputCount = 2, settings.maxInputBufferSize, this) {
+    override def onError(input: Int, e: Throwable): Unit = fail(e)
+  }
+
+  class ChoppingBlock(idx: Int, name: String) extends TransferState {
+    override def isReady: Boolean = buffer.nonEmpty
+    override def isCompleted: Boolean = false
+
+    private var buffer = ByteString.empty
+
+    /**
+     * Whether there are no bytes lying on this chopping block.
+     */
+    def isEmpty: Boolean = buffer.isEmpty
+
+    /**
+     * Pour as many bytes as are available either on the chopping block or in
+     * the inputBunch’s next ByteString into the supplied ByteBuffer, which is
+     * expected to be in “read left-overs” mode, i.e. everything between its
+     * position and limit is retained. In order to allocate a fresh ByteBuffer
+     * with these characteristics, use `prepare()`.
+     */
+    def chopInto(b: ByteBuffer): Unit = {
+      b.compact()
+      if (buffer.isEmpty) {
+        buffer = inputBunch.dequeue(idx) match {
+          case bs: ByteString ⇒ bs
+          case SendBytes(bs)  ⇒ bs
+          case n: NegotiateNewSession ⇒
+            setNewSessionParameters(n)
+            ByteString.empty
+        }
+        if (tracing) log.debug(s"chopping from new chunk of ${buffer.size} into $name (${b.position})")
+      } else {
+        if (tracing) log.debug(s"chopping from old chunk of ${buffer.size} into $name (${b.position})")
+      }
+      val copied = buffer.copyToBuffer(b)
+      buffer = buffer.drop(copied)
+      b.flip()
+    }
+
+    /**
+     * When potentially complete packet data are left after unwrap() we must
+     * put them back onto the chopping block because otherwise the pump will
+     * not know that we are runnable.
+     */
+    def putBack(b: ByteBuffer): Unit =
+      if (b.hasRemaining()) {
+        if (tracing) log.debug(s"putting back ${b.remaining} bytes into $name")
+        val bs = ByteString(b)
+        if (bs.nonEmpty) buffer = bs ++ buffer
+        prepare(b)
+      }
+
+    /**
+     * Prepare a fresh ByteBuffer for receiving a chop of data.
+     */
+    def prepare(b: ByteBuffer): Unit = {
+      b.clear()
+      b.limit(0)
```

Comment:
because then I can just always `.compact()` and it works—compacting a cleared buffer yields a completely full buffer (don’t you love that API?)